File: rxflux2/src/main/java/com/huyingbao/rxflux2/action/RxAction.java
package com.huyingbao.rxflux2.action;

import android.support.annotation.NonNull;
import android.support.v4.util.ArrayMap;

/**
 * Object class that hold the mType of action and the mData we want to attach to it
 */
public class RxAction {
    private final String mType;
    private final ArrayMap<String, Object> mData;

    RxAction(@NonNull String type, ArrayMap<String, Object> data) {
        this.mType = type;
        this.mData = data;
    }

    public String getType() {
        return mType;
    }

    public ArrayMap<String, Object> getData() {
        return mData;
    }

    @SuppressWarnings("unchecked")
    public <T> T get(String tag) {
        return (T) mData.get(tag);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (!(obj instanceof RxAction)) return false;
        RxAction rxAction = (RxAction) obj;
        if (!mType.equals(rxAction.mType)) return false;
        return !(mData != null ? !mData.equals(rxAction.mData) : rxAction.mData != null);
    }

    @Override
    public int hashCode() {
        int result = mType.hashCode();
        result = 31 * result + (mData != null ? mData.hashCode() : 0);
        return result;
    }

    @Override
    public String toString() {
        return "RxAction{" + "mType='" + mType + '\'' + ", mData=" + mData + '}';
    }

    public static class Builder {
        private String mType;
        private ArrayMap<String, Object> mData;


        public Builder(@NonNull String type) {
            this.mType = type;
            this.mData = new ArrayMap<>();
        }


        /**
         * 向RxAction中添加key-value
         *
         * @param key   不能为空
         * @param value 为空时，key和value都不添加到action中
         * @return
         */
        public Builder put(@NonNull String key, Object value) {
            if (value != null) mData.put(key, value);
            return this;
        }

        /**
         * 生成RxAction对象
         *
         * @return
         */
        public RxAction build() {
            return new RxAction(mType, mData);
        }
    }
}

File: rxflux2-kotlin/src/main/java/com/huyingbao/rxflux2/action/RxAction.kt
package com.huyingbao.rxflux2.action

import android.support.v4.util.ArrayMap

/**
 * Object class that hold the mType of action and the mData we want to attach to it
 */
open class RxAction internal constructor(val type: String, val data: ArrayMap<String, Any>?) {

    operator fun <T> get(tag: String): T {
        return data!![tag] as T
    }

    override fun equals(obj: Any?): Boolean {
        if (this === obj) return true
        if (obj !is RxAction) return false
        val rxAction = obj as RxAction?
        return if (type != rxAction!!.type) false else !if (data != null) data != rxAction.data else rxAction.data != null
    }

    override fun hashCode(): Int {
        var result = type.hashCode()
        result = 31 * result + (data?.hashCode() ?: 0)
        return result
    }

    override fun toString(): String {
        return "RxAction{mType='$type', mData=$data}"
    }

    class Builder(private val mType: String) {
        private val mData: ArrayMap<String, Any>


        init {
            this.mData = ArrayMap()
        }


        /**
         * 向RxAction中添加key-value
         *
         * @param key   不能为空
         * @param value 为空时，key和value都不添加到action中
         * @return
         */
        fun put(key: String, value: Any?): Builder {
            if (value != null) mData.put(key, value)
            return this
        }

        /**
         * 生成RxAction对象
         *
         * @return
         */
        fun build(): RxAction {
            return RxAction(mType, mData)
        }
    }
}